### PR TITLE
[Model] Rewrite GraphSAGE example to use DataLoader

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Rewrote GraphSAGE example to use DataLoader for sampling.

Profiling result goes as follows:

| # GPUs | # sampler workers per GPU | Time per epoch |
|:---:|:---:|:---:|
| 1 | 1 | 29.68s | 
| 1 | 6 | 17.2401s |
| 1 | 12 | 17.1699s |
| 4 | 1 | 7.49s |
| 4 | 6 | 7.1922s |

The benefit of multiprocessed sampling also diminishes as number of workers increase:

| # GPUs | # sampler workers per GPU | Time per epoch on sampler only |
|:---:|:---:|:---:|
| 1 | 1 | 11.5548s | 
| 1 | 6 | 3.2159s |
| 1 | 12 | 4.0606s |

EDIT (Mar 9th): the reason may be that the dataset is too small as the batch size is 1000 and we only have around 200k nodes - which yields us 200 batches.  MXNet and Tensorflow both have their own solutions analogous to PyTorch DataLoader, so I expect that supporting multiprocess sampling should be straightforward enough.

Also fixes a bug on evaluation code with multi-GPU.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
Can sampling now pipeline with the computation?

**Comment by BarclayII** [comment]
> Can sampling now pipeline with the computation?

If "pipelining" means to sample and compute simultaneously then it should be naturally achieved with `DataLoader` with `num_workers > 1`.

File: examples/pytorch/graphsage/graphsage_sampling.py [hunk 1]
```diff
@@ -194,7 +208,16 @@ def run(proc_id, n_gpus, args, devices, data):
...
+    sampler = NeighborSampler(g, [int(fanout) for fanout in args.fan_out.split(',')])
+
+    # Create PyTorch DataLoader for constructing blocks
+    dataloader = DataLoader(
+        dataset=train_nid.numpy(),
+        batch_size=args.batch_size,
+        collate_fn=sampler.sample_blocks,
+        shuffle=True,
+        drop_last=False,
+        num_workers=args.num_workers_per_gpu)
```
Comment (zheng-da) [inline examples/pytorch/graphsage/graphsage_sampling.py]: can we have builtin sampler for neighbor sampling?

Comment (BarclayII) [reply examples/pytorch/graphsage/graphsage_sampling.py]: What do you mean by builtin sampler?  Everything in C?

Comment (jermainewang) [reply examples/pytorch/graphsage/graphsage_sampling.py]: Quan and I think we could stay with the current implementation first and gradually putting some common samplers to DGL package once we have more examples built up. @zheng-da what do you think?

**Pull request merged**